### PR TITLE
Shiny Hunter for Shaymin

### PR DESCRIPTION
**Pull request by Gr33nBug**
A new shiny hunter choice done specifically for the static Shaymin event on Brilliant Diamond/Shiny Pearl. This new one takes between 30 and 35 seconds long for each try versus the almost 60 seconds of the normal static shiny hunter. 
It replicates in broad terms this [video (YouTube link)](https://youtu.be/gG1AOOxC6UY?si=OrfooZcim4G8nNhC) but now uses the bike in order to save more time.

Tested on the virtual machine recommended on the tutorial and on real hardware.

How to use this new Shiny Hunter:

1. The player must be facing Shaymin from its down side.
2. The player must have their bike as the only key item registered (Quick item menu is the one using the PLUS/START button).
3. It is imperative that the bike has its fast gear enabled (To change the bike gear the player must be riding the bike and press the B button once). To know if you are using the fast gear, tap your DPAD in one direction and the character while riding the bike will move 2 positions on the grid instead of just 1.
4. Face Shaymin, save your game and run the program as always.

----------------------------------
Thanks to @David34920 to make me reconsider using the bike.

@Dinones Could it be possible to add when the PLUS and MINUS buttons are being pressed on the GUI? Thank you in advance.

### CODE REVIEW

**Comment by David34920** [comment]
Great, I hope I will be able to use it soon :)

**Comment by Gr33nBug** [comment]
Just wanted to say that it worked. After 5459 tries between the normal static method and this one I finally have a shiny Shaymin.
I have both editions of the games so I'll probably try to get a second one.

File: Modules/Control_System.py [hunk 1]
```diff
@@ -254,6 +254,114 @@ def starter_encounter(image, state):
...
+        # Look for the text box
+        if is_text_box_visible(image):
+            return 'ESCAPE_COMBAT_1'
+
+        # Check the elapsed time
+        if image.shiny_detection_time and time() - image.shiny_detection_time >= CONST.SHINY_DETECTION_TIME:
+            return 'SHINY_FOUND'
+    
+    # Combat loaded (Both Pokémon in the field)
+    elif state == 'ESCAPE_COMBAT_1':
```
Comment (David34920) [inline Modules/Control_System.py:305]: Is it worth and possible to factorize code with the wild encounter section?

**Comment by Dinones** [comment]
> @Dinones Could it be possible to add when the PLUS and MINUS buttons are being pressed on the GUI? Thank you in advance.

Sure! I'll try my best with my professional paint experience jajaja. By the way, I have not accepted the request yet because I've been so busy and I want to review it with patience, but I will do it. Thank you for your help! 

Did you find the shiny Shaymin, by the way?

**Comment by Gr33nBug** [comment]
> Did you find the shiny Shaymin, by the way?

I did find one in Brilliant Diamond but I have not tried in Shiny Pearl yet. 

> Sure! I'll try my best with my professional Paint experience jajaja. By the way, I have not accepted the request yet because I've been so busy and I want to review it with patience, but I will do it before implementing new features. Thank you for your help!

I asked about the plus and minus buttons because while I was testing my code I thought it wasn't working because I didn't see the plus button being pressed on the GUI. The plus button needs to be pressed LONGER than an usual A press for the bike to spawn. I hope it's not a hard task to add the new button presses. I didn't do it myself as I thought you would be more knowledgeable about how it works and wouldn't take you as much time it would take me so don't worry about reviewing this soon. Focus on the features you find yourself motivated to do. 

About the changes to the macro I made recently, it's something I want to test because I just simply don't like using time.sleep() for those moments the script needs to wait.
I've been studying the NXBT project and I'm pretty sure what I explained in commit [7fbbf78](https://github.com/Dinones/Nintendo-Switch-Pokemon-Shiny-Hunter/pull/46/commits/7fbbf78a9aab6ecb268ec4b562ad45bdb1653a40) will not work and that's why I reverted it. So I want to see how specifically the pauses are resolved in a text macro. I have not found how it does it yet...
I know this goes beyond what this project is about but I just wanted to explain my efforts.

Just leave the PR open for the time being and I'll try to test the macro this weekend on my Shiny Pearl save. 

And thank YOU for this fantastic project.

File: Modules/Control_System.py (hunk 1 above)
Comment (Dinones) [reply Modules/Control_System.py:305]: Do you both think we should made a "Special Encounter" category for these cases in the selection menu? Like:

[1] Wild encounter
[2] Static encounter
[3] Starter encounter
[4] Special encounters
[X] Fishing _(future implementation)_
[X] Pokéradar _(future implementation)_
[X] Fossils _(future implementation)_
[X] Eggs _(future implementation)_

And if Special encounters are selected, print another selection menu with:

- Shaymin
- Creselia/Mesprit
- Maybe Dialga and Palkia _(automate the STATIC_ENCOUNTERS_DELAY)_
- Arceus _(automate the STATIC_ENCOUNTERS_DELAY)_
- Happiny Egg
- Etc.

**Comment by Dinones** [comment]
> I did find one in Brilliant Diamond but I have not tried in Shiny Pearl yet.

Well, should be the same. I don't think there is any difference between both versions

> I asked about the plus and minus buttons because while I was testing my code I thought it wasn't working because I didn't see the plus button being pressed on the GUI. The plus button needs to be pressed LONGER than an usual A press for the bike to spawn. I hope it's not a hard task to add the new button presses. I didn't do it myself as I thought you would be more knowledgeable about how it works and wouldn't take you as much time it would take me so don't worry about reviewing this soon. Focus on the features you find yourself motivated to do.

Oh I see, printing buttons in the GUI works independently from pressing the buttons themselves; it was more like an extra cool implementation rather than a functional one, but, as said, I'll do my best with Paint :)

> About the changes to the macro I made recently, it's something I want to test because I just simply don't like using time.sleep() for those moments the script/macro needs to wait. I've been studying the NXBT project and I'm pretty sure what I explained in commit [7fbbf78](https://github.com/Dinones/Nintendo-Switch-Pokemon-Shiny-Hunter/pull/46/commits/7fbbf78a9aab6ecb268ec4b562ad45bdb1653a40) will not work and that's why I reverted it. So I want to see how specifically the pauses are resolved in a text macro. I have not found how it does it yet... I know this goes beyond what this project is about but I just wanted to explain my efforts.

Why don't you like using time.sleep()? Macros are executed in a different thread from the main one, so there is no issue at all if they are just waiting some time. But explain to me your point of view; maybe there's something I don't see

> And thank YOU for this fantastic project.

Nope, thank you for helping me; we are getting a really cool project :)

File: Modules/Control_System.py (hunk 1 above)
Comment (Gr33nBug) [reply Modules/Control_System.py:305]: If your plans are to expand to other pokémon games on switch, at some point it's going to be necessary to have each kind of hunt inside of a submenu after choosing the game that is going to be played.

**Comment by Gr33nBug** [comment]
I leave here a video for you to see how the loop works. Ignore the frame drops, they are because using OBS on my laptop is too much.

https://github.com/user-attachments/assets/3e80ff14-c0b6-45de-928c-e7042bf9e71c

File: Modules/Control_System.py (hunk 1 above)
Comment (Dinones) [reply Modules/Control_System.py:305]: You are right, I was thinking about creating a very simple interface where users can change Constants without accessing the _Constants.py_ file, ensuring this way values and format are correct.

**Comment by Dinones** [comment]
I like how it looks, but why are down/up buttons pressed between 0:16 and 0:23? It should just wait, no?

**Comment by Gr33nBug** [comment]
I didn't notice that. It's repeating the macro present in walk_down_and_up_for_shaymin_respawn function. It has to be something related to when the state is checked like when A is constantly being pressed in Enter_Static_Combat_1 as the state has not changed yet.

It may be a good idea to split the macro in smaller pieces as to minimize the chances that it does something unexpected but it looks that the battle loading and the other checks takes longer that the repetition of the macro so we could leave it like that. 

I would like to see how this hunter performs for other people because right now it's doing other weird things like 3 false positives in the last 50 minutes and also some random soft resets. Just check the code when you have time and motivation.

**Comment by Gr33nBug** [comment]
I think I found why the walk_down_then_up macro repeats itself. The animation to get down from the bike plus the camera closeup to shaymin takes about 2-3 seconds long together. I'll extend the last sleep() in the macro and test it when I have the chance.
I'll take in consideration commit [73cf604](https://github.com/Dinones/Nintendo-Switch-Pokemon-Shiny-Hunter/commit/73cf604f8503a1300623aed90252776134eba61b) and see if it also helps with the false positives I was experiencing last week.

**Comment by Dinones** [comment]
Hi thank you for the update! After the changes you made, everything is working correctly?

**Comment by Gr33nBug** [comment]
There is nothing else I can do to improve it more at the moment. It seems to be working as fine as when I caught my shiny.
Check it, read the code and the comments I wrote, and merge it at your convenience.

**Comment by Dinones** [comment]
Hi @Gr33nBug! Thank you a lot for the implementation! As it is a big implementation, I want to dedicate an appropriate time to review all changes. I'll try to do so this weekend. Nevertheless, I'm not able to reach Shaymin to test it out, so I will just check the code, probably make some minor changes, and accept your PR.

Again, thank you!

**Comment by Dinones** [comment]
Hi @Gr33nBug! Again, Thank you a lot for the implementation! I have already merged your PR.

**Comment by Gr33nBug** [comment]
Hi @Dinones.
The player **MUST** have the bike as the only item in the PLUS menu.
If you try to assign a second, third and fourth item to that menu you are asked to select one position (up, down, left, right). When the PLUS button is pressed then you have to choose one of those positions. That would waste time.

I think the explanation in my first post is good enough and foolproof _crosses fingers_ but maybe images and a video example would be required in order to avoid future issues. 

Anyway, thank you for your time.

**Comment by Dinones** [comment]
Perfect! Don't worry, everything is clear, thank you a lot! I just wanted to be sure in order to write the instructions correctly.

**Pull request merged**